react/features/prejoin/components/country-picker/CountryPicker.js

```diff
@@ -0,0 +1,245 @@
+        this._onDropdownClose = this._onDropdownClose.bind(this);
+        this._onCountrySelectorClick = this._onCountrySelectorClick.bind(this);
+        this._onEntryClick = this._onEntryClick.bind(this);
+        this._onKeyPress = this._onKeyPress.bind(this);
+    }
+
+    _onChange: (Object) => void;
+
+    /**
+     * Handles the input text change.
+     * Automatically updates the country from the 'CountrySelector' if a
+     * phone number prefix is entered (00 or +).
+     *
+     * @param {Object} e - The synthetic event.
+     * @returns {void}
+     */
+    _onChange({ target: { value } }) {
+        if (PREFIX_REG.test(value)) {
+            const textWithDialCode = value.replace(PREFIX_REG, '');
+
+            if (textWithDialCode.length >= 4) {
+                const country = getCountryFromDialCodeText(textWithDialCode);
+
+                if (country) {
+                    const rest = textWithDialCode.replace(country.dialCode, '');
+
+                    this.props.setDialOutCountry(country);
+                    this.props.setDialOutNumber(rest);
+
+                    return;
+                }
+            }
+        }
+
+        this.props.setDialOutNumber(value);
+    }
+
+    _onCountrySelectorClick: (Object) => void;
+
+    /**
+     * Click handler for country selector.
+     *
+     * @param {Object} e - The synthetic event.
+     * @returns {void}
+     */
+    _onCountrySelectorClick(e) {
+        e.stopPropagation();
+
+        if (this.state.isOpen) {
+            return this.setState({
+                isOpen: false
+            });
+        }
+
+        this.setState({
+            isOpen: true
+        });
+    }
+
+    _onDropdownClose: () => void;
+
+    /**
+     * Closes the dropdown.
+     *
+     * @returns {void}
+     */
+    _onDropdownClose() {
+        this.setState({
+            isOpen: false
+        });
+    }
+
+    _onEntryClick: (Object) => void;
+
+    /**
+     * Click handler for a single entry from the dropdown.
+     *
+     * @param {Object} country - The country used for dialing out.
+     * @returns {void}
+     */
+    _onEntryClick(country) {
+        this.props.setDialOutCountry(country);
+        this._onDropdownClose();
+    }
+
+    _onKeyPress: (Object) => void;
+
+    /**
+     * Handler for key presses.
+     *
+     * @param {Object} e - The synthetic event.
+     * @returns {void}
+     */
+    _onKeyPress(e) {
+        if (e.key === 'Enter') {
+            this.props.onSubmit();
+        }
+    }
+
+    /**
+     * Implements React's {@link Component#componentDidUnmount()}.
+     *
+     * @inheritdoc
+     */
+    componentDidMount() {
+        this.inputRef.current.focus();
+    }
+
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     * @returns {ReactElement}
+     */
+    render() {
+        const { dialOutCountry, dialOutNumber } = this.props;
+        const { isOpen } = this.state;
+        const {
```

Comment:
We don't usually decompose `this` because it leads to confusion with imported functions. But i'm not strongly agains it.

What do you think @saghul ?

Reply:
But local methods use underscore so I don't think there is chance of collision.